File: project_euler/Problem044/Problem044.go
package project_euler

import (
    "fmt"
    "math"
)

func Problem044() {
    pentagons := make([]int, 0)
    index := 1
    for true {
        current := index * (3 * index - 1) >> 1
        for _, pentagon := range pentagons {
            if isPentagon(current + pentagon) && isPentagon(current - pentagon) {
                fmt.Println(current - pentagon)
                return
            }
        }
        pentagons = append(pentagons, current)
        index++
    }
}

func isPentagon(value int) bool {
    auxiliary := math.Sqrt(float64(value * 24 + 1))
    return math.Abs(math.Floor(auxiliary + 0.5) - auxiliary) < 0.00000001 && 0 == (int(auxiliary) + 1) % 6
}

File: project_euler/Problem044/Problem044.js
function Problem044() {
    let pentagons = []
    let index = 1
    while (true) {
        let current = index * (3 * index - 1) >>> 1
        for (let pentagon of pentagons) {
            if (isPentagon(current + pentagon) && isPentagon(current - pentagon)) {
                console.log(current - pentagon)
                return
            }
        }
        pentagons.push(current)
        ++index
    }
}

function isPentagon(value) {
    let auxiliary = Math.sqrt(value * 24 + 1)
    return Math.abs(Math.floor(auxiliary + 0.5) - auxiliary) < 0.00000001 && !((Math.floor(auxiliary) + 1) % 6)
}

module.exports = {
    Problem044 : Problem044
};
